### DISCUSSION TITLE
re: Migrate Users and Overloads?

### DISCUSSION THREAD

**Discussion by osseonews** [nhost/nhost#705] [General]
First of all, congrats on this service. This is an amazing idea. We are huge fans of Hasura, but managing the user authentication and files via Hasura, has always been a huge pain. This service makes create an App with Hasura backend much faster, as far as I can tell so far. 

2 quick questions:
1. Migration: We want to migrate our current users from the our current Hasura instance to nhost. What is the best way to insert these users into users table on nhost? We can't migrate passwords, so do we just migrate the email and leave the other fields blank? If the password field is blank on nhost, will the user be prompted to reset it upon login on the frontend, using the nhost hooks or do we need to handle that ourselves by actively calling the use-reset-password hook ? Thanks.
2. Overloads: What is the meaning of the multiple over Overloads for each hook in the NextJS documentation. each example seems to be exactly the same, so why are they repeated with Overload 1 and Overload 2?

**Comment by elitan** [comment]
Hi @osseonews

Sorry for the delayed response.

Thanks for your nice words about our service. That's exactly what we're trying to provide. Build cool stuff fast 🚀.

## Migration

Yea you can add users manually to the `auth.users` table without a password. Users could still sign in using [Magic Link](https://docs.nhost.io/platform/authentication/sign-in-with-magic-link) and set [change password](https://docs.nhost.io/reference/javascript/auth/change-password) once they are signed in.

## Overloads

The reason there are two overloads sometime in our [Next.js docs](https://docs.nhost.io/reference/nextjs) is because we changed the parameters for some hooks. 

Example: [`useChangeEmail`](https://docs.nhost.io/reference/nextjs/use-change-email).

Instead of adding the `email` to the hook the `email` is used in the `changeEmail` method that is returned from `useChangeEmail`.

Old (depricated):

```js
const { changeEmail } = useChangeEmail({ email })

...

await changeEmail();
```

New:

```js
const { changeEmail } = useChangeEmail()

...

await changeEmail({ email });
```

I hope that explains things a bit. Let me know if you want me to expand on anything.

**Comment by osseonews** [comment]
Thanks for the response. I honestly love what you guys are doing here. The market needs an outsourced Hasura solution with out of the box features, like authentication and files. Hasura's own cloud hosting solution provides little value to most biz. 

Anyway, instead of opening up a new issue, one thing I still don't quite understand is how to actually upload images from a React app to the nhost backend. The docs are scarce on this. Do you have any actual code with a React form onSubmit to upload an image? We are using Nextjs and for file uploads we use https://uppy.io/docs/react/dashboard/ . 

In general, the storage docs are quite vague and it's difficult of figure out how things work, e.g. how do create a new bucket?

**Reply by elitan** [reply]
Here's an example of how to upload a file: https://github.com/nhost/nhost/blob/main/examples/react-apollo-crm/src/components/CustomerAddComment.tsx#L18-L21.

You can create buckets manually via the Hasura by creating a new bucket in the `storage.buckets` table.

**Comment by osseonews** [comment]
I still have an issue with migration. You mention above: "Yea you can add users manually to the auth.users table without a password. "

Your docs (https://docs.nhost.io/platform/authentication/users) say: "Never create users directly via GraphQL or database."

Why? Actually, our whole plan is to just to a bulk insert using graphql of all our current users into the nhost auth.users table with a server side node application. Why would this be an issue? Why can't we create users via GraphQL? The nhost client can only be used on the client, so using the nhost.auth.sighup is not really possible on a server. Anyway, it's easier to just use a regular graphql library on the server to call the backend and insert users. 

Also the nhost dashboard now as a form to add a user. Is this also not possible to use yet? Why is it there, if the docs say to never create a user via the database?
